src/transaction/transaction.ts

```diff
@@ -1,45 +1,128 @@
 } from './types'
 import { encodeCbor } from '../util'
+import { CardanoEra } from '../types'
+import { SigningParameters } from '../crypto-providers/types'
+
+type WitnessSet = Map<number, unknown[]>
+
+// If witnessSet is empty, the CBOR lib parses it as {}, otherwise it seems to always return a Map.
+// Let's convert it to Map in any case, just to make sure.
+const _parseWitnessSet = (witnessSet: unknown): WitnessSet => {
+  const clonedWitnessSet = cloneDeep(witnessSet)
+  return (clonedWitnessSet instanceof Map)
+    ? clonedWitnessSet
+    : new Map(Object.entries(clonedWitnessSet as object)) as unknown as WitnessSet
+}
 
 const TxByronWitness = (
   publicKey: Buffer, signature: Buffer, chaincode: Buffer, addressAttributes: object,
 ): TxWitnessByron => [publicKey, signature, chaincode, encodeCbor(addressAttributes)]
 
 const TxShelleyWitness = (publicKey: Buffer, signature: Buffer): TxWitnessShelley => [publicKey, signature]
 
-const TxSigned = (
-  rawTx: RawTransaction,
+const _rawTxToTxSigned = (
+  params: SigningParameters,
   byronWitnesses: TxWitnessByron[],
   shelleyWitnesses: TxWitnessShelley[],
 ): TxCborHex => {
-  const { body, scriptWitnesses, auxiliaryData } = rawTx
-  const nativeScriptWitnessList = scriptWitnesses as any[] | undefined
-  const witnessSet = new Map()
+  const {
+    body, scriptWitnesses, datumWitnesses, redeemerWitnesses, scriptValidity, auxiliaryData,
+  } = params.rawTx!
+
+  // cardano-cli deduplicates the script witnesses before adding them to the signed transaction
+  const scriptWitnessList = uniqWith((scriptWitnesses || []) as any[], isEqual)
+  const datumWitnessList = (datumWitnesses || []) as any[]
+  const redeemerWitnessList = (redeemerWitnesses || []) as any[]
+
+  let nativeScriptWitnessList: unknown[] = []
+  let plutusScriptWitnessList: unknown[] = []
+  if (params.era === CardanoEra.ALONZO) {
+    // In alonzo era txs, each native script is packed in a list with 0 literal, each plutus script
+    // is packed in a list with 1 literal and they are all stored in scriptWitnessList.
+    // We need to partition them back and unpack the actual scripts.
+    nativeScriptWitnessList = scriptWitnessList.filter((s) => s[0] === 0).map((s) => s[1])
+    plutusScriptWitnessList = scriptWitnessList.filter((s) => s[0] === 1).map((s) => s[1])
+  } else {
+    // In older txs, scriptWitnessList is just a list of native scripts.
+    nativeScriptWitnessList = scriptWitnessList
+  }
+
+  const witnessSet: WitnessSet = new Map()
+
   if (shelleyWitnesses.length > 0) {
     witnessSet.set(TxWitnessKeys.SHELLEY, shelleyWitnesses)
   }
-  if (nativeScriptWitnessList?.length) {
-    // cardano-cli deduplicates the script witnesses before adding them to the signed transaction
-    witnessSet.set(TxWitnessKeys.NATIVE_SCRIPTS, uniqWith(nativeScriptWitnessList, isEqual))
+
+  if (nativeScriptWitnessList.length > 0) {
+    witnessSet.set(TxWitnessKeys.NATIVE_SCRIPTS, nativeScriptWitnessList)
   }
+
```

Comment:
I guess I'd move this up near shelleyWitnesses, that's where it semantically belongs and how treat them in the code (irrespective of what the enum values are).

Reply:
Moved